lib/api/src/entities/engine/inner.rs

```diff
@@ -0,0 +1,150 @@
+                "The selected runtime does not support `deserialize_unchecked`".into(),
+            )),
+        }
+    }
+
+    #[cfg(all(feature = "sys", not(target_arch = "wasm32")))]
+    /// Deserializes a WebAssembly module which was previously serialized with
+    /// `Module::serialize`,
+    ///
+    /// # Errors
+    /// Not every implementer supports serializing and deserializing modules.
+    /// Currently, only the `sys` engines support it, and only when the target
+    /// architecture is not `wasm32`.
+    pub(crate) unsafe fn deserialize(
+        &self,
+        bytes: impl IntoBytes,
+    ) -> Result<Arc<Artifact>, DeserializeError> {
+        match self {
+            #[cfg(feature = "sys")]
+            Self::Sys(s) => s.deserialize(bytes.into_bytes().to_owned().into()),
+            _ => Err(DeserializeError::Generic(
+                "The selected runtime does not support `deserialize`".into(),
+            )),
+        }
+    }
+
+    #[cfg(all(feature = "sys", not(target_arch = "wasm32")))]
+    /// Load a serialized WebAssembly module from a file and deserialize it.
+    ///
+    /// # Note
+    /// You should almost always prefer [`Self::deserialize_from_file`].
+    ///
+    /// # Errors
+    /// Not every implementer supports serializing and deserializing modules.
+    /// Currently, only the `sys` engines support it, and only when the target
+    /// architecture is not `wasm32`.
+    ///
+    /// # Safety
+    /// See [`Artifact::deserialize_unchecked`].
+    pub(crate) unsafe fn deserialize_from_file_unchecked(
+        &self,
+        file_ref: &Path,
+    ) -> Result<Arc<Artifact>, DeserializeError> {
+        match self {
+            #[cfg(feature = "sys")]
+            Self::Sys(s) => s.deserialize_from_file_unchecked(file_ref),
+            _ => Err(DeserializeError::Generic(
+                "The selected runtime does not support `deserialize_from_file_unchecked`".into(),
+            )),
+        }
+    }
+
+    #[cfg(all(feature = "sys", not(target_arch = "wasm32")))]
+    /// Load a serialized WebAssembly module from a file and deserialize it.
+    ///
+    /// # Errors
+    /// Not every implementer supports serializing and deserializing modules.
+    /// Currently, only the `sys` engines support it, and only when the target
+    /// architecture is not `wasm32`.
+    ///
+    /// # Safety
+    /// See [`Artifact::deserialize`].
+    pub(crate) unsafe fn deserialize_from_file(
+        &self,
+        file_ref: &Path,
+    ) -> Result<Arc<Artifact>, DeserializeError> {
+        match self {
+            #[cfg(feature = "sys")]
+            Self::Sys(s) => s.deserialize_from_file(file_ref),
+            _ => Err(DeserializeError::Generic(
+                "The selected runtime does not support `deserialize_from_file`".into(),
+            )),
+        }
+    }
+}
+
+impl Default for RuntimeEngine {
+    fn default() -> Self {
+        #[cfg(feature = "sys")]
+        {
```

Comment:
This code will probably give build warnings if multiple backends are enabled.

Probably either need to write that in a more if/else style, or ignore the warning.

Reply:
1005cfc6f721b42dad2e6b519d5dc0f440618780